src/main/java/jenkins/branch/MultiBranchProjectDisplayNamingTrait.java

```diff
@@ -0,0 +1,104 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.branch;
+
+import org.jenkinsci.Symbol;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Extension;
+import hudson.util.ListBoxModel;
+import jenkins.scm.api.SCMHead;
+import jenkins.scm.api.trait.SCMSourceTrait;
+import jenkins.scm.api.trait.SCMSourceTraitDescriptor;
+import jenkins.scm.impl.trait.Discovery;
+
+/**
+ * A {@link SCMSourceTrait} that controls how to set display names for {@link SCMHead}s
+ * with additional information like merge/change/pull requests.
+ *
+ * @since 2.7.1
+ */
+public class MultiBranchProjectDisplayNamingTrait extends SCMSourceTrait {
+    /**
+     * The display naming strategy.
+     */
+    @NonNull
+    private final MultiBranchProjectDisplayNamingStrategy displayNamingStrategy;
+
+    /**
+     * Constructor for both stapler and programmatic instantiation.
+     *
+     * @param displayNamingStrategy the {@link MultiBranchProjectDisplayNamingStrategy}.
+     */
+    @DataBoundConstructor
+    public MultiBranchProjectDisplayNamingTrait(MultiBranchProjectDisplayNamingStrategy displayNamingStrategy) {
+        this.displayNamingStrategy = displayNamingStrategy;
+    }
+
+    /**
+     * Gets the {@link MultiBranchProjectDisplayNamingStrategy}.
+     *
+     * @return the {@link MultiBranchProjectDisplayNamingStrategy}.
+     */
+    @NonNull
+    public MultiBranchProjectDisplayNamingStrategy getDisplayNamingStrategy() {
+        return displayNamingStrategy;
+    }
+
+    /**
+     * Our descriptor.
+     */
+    @Symbol("multiBranchProjectDisplayNaming")
+    @Extension
+    @Discovery
+    public static class DescriptorImpl extends SCMSourceTraitDescriptor {
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public String getDisplayName() {
+            return Messages.MultiBranchProjectDisplayNamingTrait_TraitDisplayName();
+        }
+
+        /**
+         * Populates the display naming strategy options.
+         *
+         * @return the display naming strategy options.
+         */
+        @NonNull
+        @Restricted(NoExternalUse.class)
+        @SuppressWarnings("unused") // stapler
+        public ListBoxModel doFillDisplayNamingStrategyItems() {
+            ListBoxModel result = new ListBoxModel();
+            result.add(Messages.MultiBranchProjectDisplayNamingTrait_Raw(), MultiBranchProjectDisplayNamingStrategy.RAW.name());
+            result.add(Messages.MultiBranchProjectDisplayNamingTrait_DisplayName(), MultiBranchProjectDisplayNamingStrategy.OBJECT_DISPLAY_NAME.name());
+            result.add(Messages.MultiBranchProjectDisplayNamingTrait_RawAndDisplayName(), MultiBranchProjectDisplayNamingStrategy.RAW_AND_OBJECT_DISPLAY_NAME.name());
+            return result;
+        }
+    }
```

Comment:
I'll try to improve this :+1:

Reply:
I switched to an enum, and I followed exactly the pattern in the other enum I found in the repository: [SuppressionStrategy](https://github.com/quilicicf/branch-api-plugin/blob/4ec3ed07be4f8f30e5444b071d5d7bc525606bd0/src/main/java/jenkins/branch/NoTriggerMultiBranchQueueDecisionHandler.java#L148-L170). It is used in 
* [this jelly template](https://github.com/quilicicf/branch-api-plugin/blob/4ec3ed07be4f8f30e5444b071d5d7bc525606bd0/src/main/resources/jenkins/branch/NoTriggerBranchProperty/config.jelly#L30-L34)
* and [this one](https://github.com/quilicicf/branch-api-plugin/blob/4ec3ed07be4f8f30e5444b071d5d7bc525606bd0/src/main/resources/jenkins/branch/NoTriggerOrganizationFolderProperty/config.jelly#L31-L35)

Since it contains the `NONE` entry, and the enum field requires a default value, I kept the `RAW` strategy for now. If you prefer removing it, please tell me what default value I should use instead :pray:

Reply:
I've asked our internal UX expert for opinions (and will update when I hear back)

in the interim my preference would be remove the `NONE` and make the `RawAndDisplayName` the default.

Reply:
I'll hold off until we hear back from the expert, I'll have plenty of work in the meantime to try and re-implement the test with a mock SCM.

Reply:
Done!